### PR TITLE
[SCFToCalyx] Add calyx::ProgramOp conversion [2/13]

### PR DESCRIPTION
**Pull request by mortbopet**
This commit introduces logic for converting a top-level MLIR 'module' into a calyx::ProgramOp.
The implementation works around an issue/restriction of the current infrastructure, being that a top-level 'module' cannot be overwritten. see: https://llvm.discourse.group/t/de-privileging-moduleop-in-translation-apis/3733/26

Furthermore, an input program is legalized using a `ConversionTarget`.

### CODE REVIEW

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+  LogicalResult mainFuncIsDefined(mlir::ModuleOp moduleOp,
+                                  StringRef topLevelFunction) {
+    if (SymbolTable::lookupSymbolIn(moduleOp, topLevelFunction) == nullptr) {
+      moduleOp.emitError("Main function '" + topLevelFunction +
+                         "' not found in module.");
+      return failure();
+    }
+    return success();
+  }
```
[the remark this answers is not in the document]
Comment (rachitnigam) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:83]: The Calyx compiler can take a command-line option (`-t`, `--toplevel`) to determine a non-standard entrypoint.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  //// Creates a new Calyx program with the contents of the source module
+  /// inlined within.
+  /// Furthermore, this function performs validation on the input function, to
+  /// ensure that we've implemented the capabilities necessary to convert it.
+  ///
+  /// @todo: this seems unnecessarily complicated:
+  /// A restriction of the current infrastructure is that a top-level 'module'
+  /// cannot be overwritten (even though this is essentially what is going on
+  /// when replacing standard::ModuleOp with calyx::ProgramOp). see:
+  /// https://llvm.discourse.group/t/de-privileging-moduleop-in-translation-apis/3733/26
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:94]: I'm not sure we are restricted here. Passes cannot adjust the IR that is "above" the operation they match, but shouldn't a ModuleOp pass be able to replace the ModuleOp with a ProgramOp?

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    return applyPartialConversion(getOperation(), target, std::move(patterns));
+  }
+
+private:
+  mlir::ModuleOp m_moduleOp;
+  calyx::ProgramOp m_programOp;
+};
+
+void SCFToCalyxPass::runOnOperation() {
+  m_moduleOp = getOperation();
```
Comment (mikeurbach) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:139]: I don't think this is being used?

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    RewritePatternSet patterns(&getContext());
+    patterns.add<ModuleOpConversion>(&getContext(), programOpOut);
+    auto target = createModuleConvTarget();
+    return applyPartialConversion(getOperation(), target, std::move(patterns));
+  }
+
+private:
+  mlir::ModuleOp m_moduleOp;
+  calyx::ProgramOp m_programOp;
```
Comment (mikeurbach) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:135]: Does this state abide by the restrictions here? https://mlir.llvm.org/docs/PassManagement/#operation-pass

More generally, what is this needed for? My understanding is to force the pass to only support converting one module in the input IR? Could we just remove this and the checks based on it? Generally, I'd expect 1 module in the input to this pass, but if there are N modules, is there a problem converting each one to its own program?

**Review by mikeurbach — COMMENTED** [review]
Some minor feedback, that can hopefully simplify this a bit. Also, is it possible to write a simple test at this point? If so, please do.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mortbopet) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:94]: > but shouldn't a ModuleOp pass be able to replace the ModuleOp with a ProgramOp?

It _should_ be possible, indeed, but as far as i was able to figure out (and as far as i am able to parse from the referenced discourse thread), that is not possible with the current way the conversion infrastructure works.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cgyurgyik) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:83]: I've opened an issue for this #1818. IMO the easiest route is to just use the brittle name comparison to `"main"`, and mention this in #1818. Would fixing this at a later point require a massive refactoring?

[the remark this answers is not in the document]
Comment (cgyurgyik) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:83]: Fixed in #1819. You should be able to get the necessary information using:

```mlir
ProgramOp program;
StringAttr topLevelName = program.entryPointName();
ComponentOp entryPointComponent = program.getEntryPointComponent();
```

after this is upstreamed. Feel free to request more functionality if this doesn't satisfy your requirements.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp (hunk 4 above)
Comment (mortbopet) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:135]: The `m_programOp` will be used to initialize the `ProgramLoweringState` of #1814. There is no real need to keep it as a member variable, so i'll modify that to a temporary variable.

Based on the criteria of the referenced site, the pass does abide by what is laid out there.

**Comment by mortbopet** [comment]
@mikeurbach Writing a test is not yet possible; for a calyx program to verify, there must be a component (named `main`) within the `calyx.ProgramOp`.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikeurbach) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:94]: I remember that thread, but the original issue being discussed was related to the translation APIs requiring a ModuleOp. Then the thread sort of digresses and starts talking about top-level ops in general. Is there a specific issue with Dialect Conversion? Does a pattern that matches  a ModuleOp and replaces it with a ProgramOp not work? If there's something preventing this from working with Dialect Conversion, perhaps we can use `applyOpPatternsAndFold` or some other pattern rewrite? It's not the end of the world, but this dynamic legality seems a little complicated for a pass that does a 1-1 replacement of the op it targets.

**Comment by mortbopet** [comment]
@mikeurbach pushed a modification to the program conversion which gets rid of the funky stuff that was being done now - `applyOpPatternsAndFold` seems to work. On the other hand, to run the conversion legalizer _some_ pattern must be provided. I've added a dummy pattern to facilitate this - please let me know if you have a better way of triggering this. The dialect conversion `OperationLegalizer` is unfortunately private to `DialectConversion`, which is why we need to go through one of the conversion driver functions.

**Comment by mikeurbach** [comment]
I see, so now the actual IR rewriting happens through `applyOpPatternsAndFold`, which looks great to me. After this change, the pass still uses DialectConversion to enforce the input pre-conditions, but doesn't rewrite anything with DialectConversion.

I think using DialectConversion's legality is a neat way to express the pre-conditions, but it makes sense that might require a dummy pattern to drive the infrastructure. I guess the alternative is to just traverse the IR and use `isa<Operation>` and `isa<Dialect>` calls.

I just took a look at a similar pass from npcomp, which enforces a pipeline's pre-conditions, and it uses DialectConversion as well: https://github.com/llvm/mlir-npcomp/blob/main/lib/Backend/Common/VerifyBackendContract.cpp. That seemed to work with an empty rewrite pattern set, so perhaps the type converter was enough to drive the infrastructure in that case. I like this approach, so whatever way you think is best to make it work here is good by me.

**Comment by mortbopet** [comment]
@cgyurgyik reworked setting of the top-level function, let me know if looks good to you.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 5]
```diff
@@ -21,15 +25,148 @@ using namespace mlir;
...
+                             << "' not found in module.";
+        return failure();
+      }
+      topLevelFunction = topLevelFunctionOpt;
+    } else {
+      /// No top level function set; infer top level if the module only contains
+      /// a single function, else, throw error.
+      auto funcOps = moduleOp.getOps<mlir::FuncOp>();
+      if (std::distance(funcOps.begin(), funcOps.end()) == 1)
+        topLevelFunction = (*funcOps.begin()).sym_name().str();
```
Comment (cgyurgyik) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:91]: Nit: May be wrong, but possibly could be simplified to `.front().sym_name().str()`.

File: include/circt/Conversion/Passes.td [hunk 6]
```diff
@@ -26,8 +26,9 @@ def SCFToCalyx : Pass<"lower-scf-to-calyx", "mlir::ModuleOp"> {
   let dependentDialects = [
     "calyx::CalyxDialect", "::mlir::scf::SCFDialect", "hw::HWDialect", "comb::CombDialect"];
   let options =
-      [Option<"topLevelComponent", "top-level-component", "std::string", "",
-              "Identifier of top-level function to be the main component">];
+      [Option<"topLevelFunctionOpt", "top-level-function", "std::string", "",
+              "Identifier of top-level function to be the main component "
```
Comment (cgyurgyik) [inline include/circt/Conversion/Passes.td:30]: Prefer `entry-point` to `main` to remain consistent with ProgramOp :-)

**Review by cgyurgyik — APPROVED** [review]
Some useless nits. Thanks for the changes.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp (hunk 5 above)
Comment (mortbopet) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:91]: `::front` is unfortunately not available for `iterator_range` :(.

**Pull request merged**